packages/ckeditor5-link/tests/autolink.js

```diff
@@ -31,6 +33,138 @@ describe( 'AutoLink', () => {
+
+			it( 'paste link', () => {
+				pasteText( 'http://hello.com' );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some [<$text linkHref="http://hello.com">selected</$text>] text</paragraph>'
+				);
+			} );
+
+			it( 'paste text including a link', () => {
+				pasteText( ' http://hello.com' );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some  http://hello.com[] text</paragraph>'
+				);
+			} );
+
+			it( 'paste not a link', () => {
+				pasteText( 'hello' );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some hello[] text</paragraph>'
+				);
+			} );
+
+			it( 'paste a HTML link', () => {
+				pasteData( {
+					'text/plain': 'http://hello.com',
+					'text/html': '<a href="http://hello.com">http://hello.com</a>'
+				} );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some [<$text linkHref="http://hello.com">selected</$text>] text</paragraph>'
+				);
+			} );
+
+			it( 'paste a HTML link that doesn not look like a link', () => {
+				pasteData( {
+					'text/plain': 'hello.com',
+					'text/html': '<a href="http://hello.com">hello.com</a>'
+				} );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some <$text linkHref="http://hello.com">hello.com</$text>[] text</paragraph>'
+				);
+			} );
+		} );
+
+		describe( 'pasting on collapsed selection', () => {
+			beforeEach( () => setData( model, '<paragraph>some [] text</paragraph>' ) );
+
+			it( 'paste link', () => {
+				pasteText( 'http://hello.com' );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some http://hello.com[] text</paragraph>'
+				);
+			} );
+		} );
+
+		describe( 'pasting on a link', () => {
+			it( 'paste with entire link selected', () => {
+				setData( model, '<paragraph>some [<$text linkHref="http://hello.com">selected</$text>] text</paragraph>' );
+				pasteText( 'http://world.com' );
+				expect( getData( model ) ).to.equal(
+					'<paragraph>some [<$text linkHref="http://world.com">selected</$text>] text</paragraph>'
+				);
+			} );
+
+			it( 'paste with partially selected link updates the entire link', () => {
+				setData( model, '<paragraph><$text linkHref="http://hello.com">some [selected] text</$text></paragraph>' );
+				pasteText( 'http://world.com' );
+				expect( getData( model ) ).to.equal(
+					'<paragraph><$text linkHref="http://world.com">some [selected] text</$text></paragraph>'
+				);
+			} );
+
+			it( 'paste with selection overlapping the start of the link inserts a new link', () => {
+				setData( model, '<paragraph>[some <$text linkHref="http://hello.com">selected] text</$text></paragraph>' );
+				pasteText( 'http://world.com' );
+				expect( getData( model ) ).to.equal(
+					// eslint-disable-next-line max-len
+					'<paragraph>[<$text linkHref="http://world.com">some selected</$text>]<$text linkHref="http://hello.com"> text</$text></paragraph>'
+				);
+			} );
+
+			it( 'paste with selection overlapping the end of the link pastes normally', () => {
```

Comment:
After giving it a thought, I don't have a clear opinion which is better: if selection and link intersects, whether autolink should be applied only on selection, or on selection+link. I also thing this is quite an edge case and people usually don't want to have two different links next to each other.

However, I think that:

1.  It should be consistent on left and right side.
2.  It should always paste a link if the selection was not collapsed.

I propose that autolinking always pastes link over the selected text, except if the selection is wholly inside the link, then we replace the whole link.

Reply:
The idea makes sense, and simplifies the code, but there's a flaw. `selection.getAttribute( 'linkHref' )` returns nothing if the selection overlaps the start of the link - it seems to be pulling attributes only from the selection start.

You can see this by opening the standard link dialog in intersection scenarios; with only end intersection the link dialog is editing the link, but with only start intersection it's inserting a new link.

For now I've left the test as inserting a new link on the left side, but if you can help me find a way to detect that scenario I'm happy to make it extend the link as well.

Reply:
> You've contradicted yourself there 🤔

I know, but I wasn't sure if I am right or not with having two links next to each other. And when not sure, I guess it is better to lean towards simpler solution. But I am as much okay with replacing the whole link + extra selection.

> The idea makes sense, and simplifies the code, but there's a flaw. `selection.getAttribute( 'linkHref' )` returns nothing if the selection overlaps the start of the link - it seems to be pulling attributes only from the selection start.

Yeah, that's correct. We do it for optimization purposes (so we won't have to look through the whole selection looking for a link). That's why we base selection state and UI state on the selection start.

In this case though, we could simply check `selection.getLastPosition()` and whether it is inside a text node with `linkHref` attribute (`position.textNode` and `node.getAttribute()`).

It should also work well for:

```
Ab<a href="#0>cx[yz</a> Foo<a href="#1">bar</a> <a href="#2">abc]xyz</a>
```

Which mixes: link on left, text, link inside and link on right.

Reply:
I'll start by adding checks for having start _or_ end in a link so the existing test scenarios work.

Supporting start _and_ end in different links seems like a much bigger change, but I'll give it a go!

Reply:
It works! My code passes the new tests, but I'm not sure if it's up to your coding standard. To keep things short I used a lot of `?.` and potentially join the selection with itself to avoid an explicit null check.